src/Microsoft.ML.FastTree/RegressionTree.cs

```diff
@@ -0,0 +1,229 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.ML.Trainers.FastTree.Internal;
+
+namespace Microsoft.ML.FastTree
+{
+    /// <summary>
+    /// A container base class for exposing <see cref="InternalRegressionTree"/>'s and
+    /// <see cref="InternalQuantileRegressionTree"/>'s attributes to users.
+    /// This class should not be mutable, so it contains a lot of read-only members.
+    /// </summary>
+    public abstract class RegressionTreeBase
+    {
+        /// <summary>
+        /// <see cref="RegressionTreeBase"/> is an immutable wrapper over <see cref="_tree"/> for exposing some tree's
+        /// attribute to users.
+        /// </summary>
+        private readonly InternalRegressionTree _tree;
+
+        /// <summary>
+        /// See <see cref="LteChild"/>.
+        /// </summary>
+        private readonly ImmutableArray<int> _lteChild;
+        /// <summary>
+        /// See <see cref="GtChild"/>.
+        /// </summary>
+        private readonly ImmutableArray<int> _gtChild;
+        /// <summary>
+        /// See <see cref="NumericalSplitFeatureIndexes"/>.
+        /// </summary>
+        private readonly ImmutableArray<int> _numericalSplitFeatureIndexes;
+        /// <summary>
+        /// See <see cref="NumericalSplitThresholds"/>.
+        /// </summary>
+        private readonly ImmutableArray<float> _numericalSplitThresholds;
+        /// <summary>
+        /// See <see cref="CategoricalSplitFlags"/>.
+        /// </summary>
+        private readonly ImmutableArray<bool> _categoricalSplitFlags;
+        /// <summary>
+        /// See <see cref="LeafValues"/>.
+        /// </summary>
+        private readonly ImmutableArray<double> _leafValues;
+
+        /// <summary>
+        /// <see cref="LteChild"/>[i] is the i-th node's child index used when
+        /// (1) the numerical feature indexed by <see cref="NumericalSplitFeatureIndexes"/>[i] is less than the
+        /// threshold <see cref="NumericalSplitThresholds"/>[i], or
+        /// (2) the categorical features indexed by <see cref="GetCategoricalCategoricalSplitFeatureRangeAt(int)"/>'s
+        /// returned value with nodeIndex=i is NOT a sub-set of <see cref="GetCategoricalSplitFeaturesAt(int)"/> with
+        /// nodeIndex=i.
+        /// Note that the case (1) happens only when <see cref="CategoricalSplitFlags"/>[i] is true and otherwise (2)
+        /// occurs. A non-negative returned value means a node (i.e., not a leaf); for example, 2 means the 3rd node in
+        /// the underlying <see cref="_tree"/>. A negative returned value means a leaf; for example, -1 stands for the
```

Comment:
>negative [](start = 50, length = 8)

Might help to use the word bitwise inverse (e.g., `~`) just for the benefit of people that know what that is. But we can refine the documentation later.